### PR TITLE
Update the README

### PR DESCRIPTION
**Pull request by karellm**
@dsifford Still a WIP but it updates the documentation. 

- Update the documentation
  - [x] Readme
  - [ ] Blog
- Small formatting and message fixes to `run.sh`:
  - [x] Fix indentations and limit line to a maximum of 80 char as much as possible
  - [x] Make code blocks more consistent (remove spaces and add visual sections)

1/ Importing and exporting the DB could use the `SEARCH_REPLACE` feature, making the command to run long and ugly. It might be worth creating an alias somewhere (or script file).

2/ You didn't keep the gulp part. Any reason for that?

### CODE REVIEW

**Comment by dsifford** [comment]
We use gulp for css/js minification / autoprefixing, but I've found that gulp is sort of a nightmare when it comes to actually doing shell-related tasks.

We currently use a a file that I called `syncup.sh` to handle the various build tasks not handled with the dockerfile (see below). Another option for the less shell-saavy would be to use npm run commands for those types of things. Example: `npm run get-database` could run a giant shell command separated by `&&`.

Here's our `syncup.sh` file

```sh
#!/bin/bash

###
# COMANDS
# 'get uploads' = retrieve all uploads that do not already exist
# 'get plugin [plugin-name]' = retrieve and overwrite the plugin [plugin-name]
# 'get --all-plugins' = retrieve all paid plugins (REQUIRED BEFORE RUNNING ENVIRONMENT)
# 'get database' = retrieve a copy of the database from the server + rename urls
# 'update plugin [plugin-name] = update a specific plugin'

# Globals
SCRIPTDIR=$(dirname "$0")
PaidPlugins=(learndash-propanel sfwd-lms um-profile-completeness)

# OS-specific variables
if [[ $(uname) == "Linux" ]]; then
  ONLINUX=true
  CONTAINER_IP="localhost"
else
  ONLINUX=false
  CONTAINER_IP=$(docker-machine ip "$DOCKER_MACHINE_NAME")
fi

cd "$SCRIPTDIR" || exit

case "$1" in

    get)

        case "$2" in

            uploads)

                cd "$SCRIPTDIR/wp-content" || exit
                echo "=> Downloading uploads from server..."
                wget -rq -nH --cut-dirs=1 -l inf --no-clobber ftp://ftp.SITEURL.com/wp-content/uploads/ --ftp-user=<USERNAME-HERE> --ftp-password=<PASSWORD-HERE>
                echo "=> Uploads Synced Successfully!"
                ;;

            --all-plugins)

                cd "$SCRIPTDIR/wp-content/plugins" || exit
                for plugin in "${PaidPlugins[@]}"; do
                    echo "=> Retrieving plugin: $plugin"
                    wget -m -nH --cut-dirs=2 ftp://ftp.SITEURL.com/wp-content/plugins/"$plugin" --ftp-user=<USERNAME-HERE> --ftp-password=<PASSWORD-HERE>
                    echo "=> Plugin $plugin Retrieved Successfully!"
                done
                ;;

            plugin)

                cd "$SCRIPTDIR/wp-content/plugins" || exit
                echo "=> Retrieving plugin: $3"
                wget -m -nH --cut-dirs=2 ftp://ftp.SITEURL.com/wp-content/plugins/"$3" --ftp-user=<USERNAME-HERE> --ftp-password=<PASSWORD-HERE>
                echo "=> Plugin $3 Retrieved Successfully!"
                ;;

            database)

                # Ensure we are in the script's directory
                cd "$SCRIPTDIR/data" || exit

                # Ensure proper file permissions
                chmod 400 aliemu_dsa

                # Run ssh-agent and add the ssh private key
                eval "$(ssh-agent -s)"
                ssh-add aliemu_dsa

                # SSH into siteground and backup the database and copy to the data directory
                if [[ $ONLINUX == true ]]; then
                    ssh -i aliemu_dsa sshuser@sshsite.net -p 18765 "
                    cd public_html
                    wp db export database.sql"
                    scp -i aliemu_dsa -P 18765 sshuser@sshsite.net:public_html/database.sql .
                else
                    ssh -i aliemu_dsa sshuser@sshsite.net -p 18765 -o "PubkeyAcceptedKeyTypes ssh-dss" "
                    cd public_html
                    wp db export database.sql"
                    scp -i aliemu_dsa -P 18765 -o "PubkeyAcceptedKeyTypes ssh-dss" sshuser@sshsite.net:public_html/database.sql .

                    # Machine not running? Start it up!
                    if [[ $(docker-machine status "$DOCKER_MACHINE_NAME") != 'Running' ]]; then
                        cd "$SCRIPTDIR" || exit
                        docker-compose up -d
                    fi
                fi

                # Import the database
                echo "=> Importing database..."
                docker exec -it "$(docker ps -lq)" bash -c "wp db import /data/database.sql --allow-root"

                # Replace live URL with dev URL
                echo "=> Replacing URLs..."
                docker exec -it "$(docker ps -lq)" bash -c "wp search-replace  'http://www.SITEURL.com' 'http://${CONTAINER_IP}:8080' --skip-columns=guid --allow-root"
                echo "=> Database Successfully Imported!"
                ;;

            *)
                echo "'get' subcommmand must be either 'uploads', 'plugin', or 'database'"
                exit 1
        esac
        ;;

    update)

        case "$2" in

            plugin)

                echo "=> Updating plugin: $3"
                PLUGIN=$3
                docker exec -it "$(docker ps -lq)" bash -c "wp plugin install $PLUGIN --force --activate --allow-root"
                ;;

            *)
                echo "'update' subcommand must be 'plugin'"
                exit 1
        esac
        ;;

    *)
        echo "The only commands available are 'get' and 'update'"
        exit 1
esac
```

From the root directory of our project, running `./syncup.sh get database` grabs a new copy of the site's database, uploads it to the container, and replaces the URLs.

**Comment by karellm** [comment]
Interesting, I actually don't like gulp (or build tools) but used it to move fast since I knew it well. Using npm could definitely be a good alternative. I will review your script and consider options.

**Comment by karellm** [comment]
@dsifford If made a cleanup pass and realized I can run `docker-compose up` with the new `run.sh`. It complains about `tput: No value for $TERM and no -T specified` and all commands fail. I need to investigate that more.

Let me know what you think of the changes too.

Finally, couple things I noticed:

1. Plugins are installed every time, themes not. Also relying on checking the folder akismet is not reliable since someone might want to use that plugin. Maybe `hello` would be a better pick.
2. We might want a script to run commands like backup/restore the DB. I like your `syncup.sh` and we might be able to reuse that

**Comment by dsifford** [comment]
@karellm Is the branch you are running from using the latest version of `run.sh`? I believe I added the `-T` flags to the tput commands.

If so, we can just remove the prettiness altogether if tput is going to cause a problem.

For themes being installed every time: The concern I had there was if someone made changes to the theme's style or options. Would those get overwritten every time the environment is ran?

Re scripts: I can write up a started `package.json` if using npm run commands was something that perhaps interested you. Otherwise, open to suggestions.

**Comment by karellm** [comment]
So the `-T` helped but didn't solve the underlying error. If I run `sudo -u www-data wp core config` inside a `docker-compose run wordpress bash`, I get `PHP Warning:  file_put_contents(/app/wp-config.php): failed to open stream: Permission denied in phar:///usr/local/bin/wp/php/commands/core.php on line 347`.

It is clearly a file/folder permission issue. Do you have any idea why?

**Comment by dsifford** [comment]
That happened to me as well! That's the strange issue that I was talking about that happened when I wasn't at my usual computer. I can't reproduce the issue from here.

The issue is: For some reason, wordpress core isn't being downloaded into the `/app` directory. So everything from then on fails (because the files don't exist).

**Comment by dsifford** [comment]
if you run `docker exec -it $(docker ps -lq) /bin/bash` and then `ls` you'll see the `/app` directory is empty. Not sure why.

**Comment by dsifford** [comment]
If you run `wp core download` before `wp core config` everything will work.

**Comment by dsifford** [comment]
I suppose we could try moving the download line into the `run.sh` script and see if that helps?

**Comment by dsifford** [comment]
Actually, standby for a bit.. I think we'd be better off grabbing wordpress manually using curl in the Dockerfile. Give me a few and I'll add that to my current PR.

**Comment by karellm** [comment]
I do have the files, I don't think it is the issue. The issue is the permissions on the folder I think.

That said, I believe downloading wordpress should be the role of the `run.sh` rather than the image. Otherwise the image will potentially get an outdated version of wordpress. And I don't plan on deploying a new one at every wordpress release.

**Comment by karellm** [comment]
So mounting the `data:/app` line in the docker file is messing with the folder permissions.

```
# without the line
drwxr-xr-x   6 www-data www-data 4096 Mar 17 19:16 app

# with the line
drwxr-xr-x   6  999   999 4096 Mar 17 19:15 app
```

**Comment by dsifford** [comment]
@karellm Why is data being mounted into app? That's a typo. Should be mounted into `/data`

Also, check my pull. I moved the wordpress install into run.sh

**Comment by karellm** [comment]
My initial intention was to persist file changes across containers but this shouldn't be needed.

**Comment by dsifford** [comment]
I see. Yeah, not needed. By creating a volume, you'll persist changes, no matter where the location might be.

**Comment by dsifford** [comment]
I guess I should also mention, if you aren't trashing your compose builds that have volumes by using `docker-compose down -v` then your `/var` directory is probably jam-packed with old volumes. 

(I made that mistake in the past and it took up a huge amount of space)

Same goes for `docker rm -v <image-ID>` :+1:

**Comment by karellm** [comment]
I have a bunch of routines to stop and rm containers and volumes. Thanks for the tip :+1: 

@dsifford See the latest change in this PR, all should be good. If you are happy with the changes, let's merge!

File: README.md [hunk 1]
```diff
@@ -29,6 +33,94 @@ We wrote a series of articles to document the project:
...
+      - /var/lib/mysql
+  wordpress:
+    image: visiblevc/wordpress:latest
+    links:
+      - db
+    ports:
+      - 8080:80
+      - 443:443
+    volumes:
+      - data:/app
```
Comment (dsifford) [inline README.md]: Double check this to make sure it matches the current `docker-compose.yml` file

**Comment by dsifford** [comment]
@karellm Other than my comment, everything looks great! Nice work :+1:

**Pull request merged**